[SPARK-8406] [SQL] Adding UUID to output file name to avoid accidental overwriting

liancheng opened the pull request:
This PR fixes a Parquet output file name collision bug which may cause data loss.  Changes made:
1.  Identify each write job issued by `InsertIntoHadoopFsRelation` with a UUID
   
   All concrete data sources which extend `HadoopFsRelation` (Parquet and ORC for now) must use this UUID to generate task output file path to avoid name collision.
2.  Make `TestHive` use a local mode `SparkContext` with 32 threads to increase parallelism
   
   The major reason for this is that, the original parallelism of 2 is too low to reproduce the data loss issue.  Also, higher concurrency may potentially caught more concurrency bugs during testing phase. (It did help us spotted SPARK-8501.)
3. `OrcSourceSuite` was updated to workaround SPARK-8501, which we detected along the way.

NOTE: This PR is made a little bit more complicated than expected because we hit two other bugs on the way and have to work them around. See [SPARK-8501](https://github.com/liancheng/spark/tree/spark-8501) and [SPARK-8513](https://github.com/liancheng/spark/tree/spark-8513).

---

Some background and a summary of offline discussion with @yhuai about this issue for better understanding:

In 1.4.0, we added `HadoopFsRelation` to abstract partition support of all data sources that are based on Hadoop `FileSystem` interface.  Specifically, this makes partition discovery, partition pruning, and writing dynamic partitions for data sources much easier.

To support appending, the Parquet data source tries to find out the max part number of part-files in the destination directory (i.e., `<id>` in output file name `part-r-<id>.gz.parquet`) at the beginning of the write job.  In 1.3.0, this step happens on driver side before any files are written.  However, in 1.4.0, this is moved to task side.  Unfortunately, for tasks scheduled later, they may see wrong max part number generated of files newly written by other finished tasks within the same job.  This actually causes a race condition.  In most cases, this only causes nonconsecutive part numbers in output file names.  But when the DataFrame contains thousands of RDD partitions, it's likely that two tasks may choose the same part number, then one of them gets overwritten by the other.

Before `HadoopFsRelation`, Spark SQL already supports appending data to Hive tables.  From a user's perspective, these two look similar.  However, they differ a lot internally.  When data are inserted into Hive tables via Spark SQL, `InsertIntoHiveTable` simulates Hive's behaviors:
1.  Write data to a temporary location
2.  Move data in the temporary location to the final destination location using
   -   `Hive.loadTable()` for non-partitioned table
   -   `Hive.loadPartition()` for static partitions
   -   `Hive.loadDynamicPartitions()` for dynamic partitions

The important part is that, `Hive.copyFiles()` is invoked in step 2 to move the data to the destination directory (I found the name is kinda confusing since no "copying" occurs here, we are just moving and renaming stuff).  If a file in the source directory and another file in the destination directory happen to have the same name, say `part-r-00001.parquet`, the former is moved to the destination directory and renamed with a `_copy_N` postfix (`part-r-00001_copy_1.parquet`).  That's how Hive handles appending and avoids name collision between different write jobs.

Some alternatives fixes considered for this issue:
1.  Use a similar approach as Hive
   
   This approach is not preferred in Spark 1.4.0 mainly because file metadata operations in S3 tend to be slow, especially for tables with lots of file and/or partitions.  That's why `InsertIntoHadoopFsRelation` just inserts to destination directory directly, and is often used together with `DirectParquetOutputCommitter` to reduce latency when working with S3.  This means, we don't have the chance to do renaming, and must avoid name collision from the very beginning.
2.  Same as 1.3, just move max part number detection back to driver side
   
   This isn't doable because unlike 1.3, 1.4 also takes dynamic partitioning into account.  When inserting into dynamic partitions, we don't know which partition directories will be touched on driver side before issuing the write job.  Checking all partition directories is simply too expensive for tables with thousands of partitions.
3.  Add extra component to output file names to avoid name collision
   
   This seems to be the only reasonable solution for now.  To be more specific, we need a JOB level unique identifier to identify all write jobs issued by `InsertIntoHadoopFile`.  Notice that TASK level unique identifiers can NOT be used.  Because in this way a speculative task will write to a different output file from the original task.  If both tasks succeed, duplicate output will be left behind.  Currently, the ORC data source adds `System.currentTimeMillis` to the output file name for uniqueness.  This doesn't work because of exactly the same reason.
   
   That's why this PR adds a job level random UUID in `BaseWriterContainer` (which is used by `InsertIntoHadoopFsRelation` to issue write jobs).  The drawback is that record order is not preserved any more (output files of a later job may be listed before those of a earlier job).  However, we never promise to preserve record order when writing data, and Hive doesn't promise this either because the `_copy_N` trick breaks the order.

liancheng committed:
Adding UUID to output file name to avoid accidental overwriting

liancheng committed:
Lets TestHive use all cores by default

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

liancheng commented:
Background and alternative solutions for this issue can be a little bit complex. Will give a summary of offline discussion with @yhuai here later.

SparkQA commented:
[Test build #35066 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35066/consoleFull) for   PR 6864 at commit [`e5e92f3`](https://github.com/apache/spark/commit/e5e92f31b86d1975c7cc69abf2421810b3f788e1).

SparkQA commented:
[Test build #35066 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35066/console) for   PR 6864 at commit [`e5e92f3`](https://github.com/apache/spark/commit/e5e92f31b86d1975c7cc69abf2421810b3f788e1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

marmbrus commented on sql/hive/src/test/scala/org/apache/spark/sql/sources/hadoopFsRelationSuites.scala:
```diff
@@ -470,6 +470,33 @@ abstract class HadoopFsRelationTest extends QueryTest with SQLTestUtils {
+  // more cores, the issue can be reproduced steadily.  Fortunately our Jenkins builder meets this
+  // requirement.  We probably want to move this test case to spark-integration-tests or spark-perf
+  // later.
+  test("SPARK-8406") {
```
Can you add a description in addition to the JIRA?

liancheng commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

liancheng commented:
The last build failure looks pretty weird: a large part of Jenkins build log output are replaced by tens of thousands of lines of integer triples, and none of the 5 test failure can be reproduced locally.

SparkQA commented:
[Test build #35077 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35077/consoleFull) for   PR 6864 at commit [`e5e92f3`](https://github.com/apache/spark/commit/e5e92f31b86d1975c7cc69abf2421810b3f788e1).

liancheng commented:
OK, found out that those integers are printed by `SQLQuerySuite.test script transform for stderr`. See [Josh's comment] [1].

SparkQA commented:
[Test build #35077 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35077/console) for   PR 6864 at commit [`e5e92f3`](https://github.com/apache/spark/commit/e5e92f31b86d1975c7cc69abf2421810b3f788e1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

liancheng commented:
Some background and a summary of offline discussion with @yhuai about this issue:

In 1.4.0, we added `HadoopFsRelation` to abstract partition support of all data sources that are based on Hadoop `FileSystem` interface.  Specifically, this makes partition discovery, partition pruning, and writing dynamic partitions for data sources much easier.  From users' perspective, what the write path does is very similar to Hive.  However, they differ a lot internally.

When data are inserted into Hive tables via Spark SQL, `InsertIntoHiveTable` simulates Hive's behaviors:

1.  Write data to a temporary location
2.  Commit the write job
3.  Move data in the temporary location to the final destination location using

    -   `Hive.loadTable()` for non-partitioned table
    -   `Hive.loadPartition()` for static partitions
    -   `Hive.loadDynamicPartitions()` for dynamic partitions

The important part is that, for appending data to existing tables in step 3, `Hive.copyFiles()` is invoked to move the data (I found the name is kinda confusing since no "copying" occurs here, we are just moving and renaming stuff).  If a file in the source directory and another file in the destination directory happen to have the same name, say `part-r-00001.parquet`, the former is moved to the destination directory and renamed with a `_copy_N` postfix (`part-r-00001_copy_1.parquet`).  That's how Hive avoids name collision.

Some alternatives fixes considered:

1.  Use similar approach as Hive

    This approach is not preferred in Spark 1.4.0 mainly because file metadata operations in S3 tend to be slow, especially for tables with lots of file and/or partitions.  That's why `InsertIntoHadoopFsRelation` just inserts to destination directory directly, and is often used together with `DirectParquetOutputCommitter` to reduce latency when working with S3.  This means, we don't have the chance to do renaming, and must avoid name collision from the beginning.

2.  Same as 1.3, just move max part number detection back to driver side

    This isn't doable because unlike 1.3, 1.4 also takes dynamic partitioning into account.  When inserting into dynamic partitions, we don't know which partition directories will be touched on driver side before issuing the write job.  Checking all partition directories is simply too expensive for tables with thousands of partitions.

3.  Add extra component to output file names to avoid name collision

    This seems to be the only reasonable solution for now.

    Currently, the ORC data source adds `System.currentTimeMillis` to the output file name.  This is not 100% safe, but only fails when two tasks with the same task ID (which implies they belong to two separate concurrent jobs) are writing to the same location within a same millisecond, which is relatively unlikely to happen.  The benefit of using a time stamp here is that, record order can be preserved.

    Another quite obvious choice is to add a UUID to the output file name.  Obviously, the benefit is this practically avoids name collision. The drawback is that record order is not preserved any more.

    However, we never promise to preserve record order when writing data, and Hive doesn't promise this either (the `_copy_N` trick breaks record order).

To sum up, adding a UUID to the output file name seems to be the simplest and safest way to fix this issue.

chenghao-intel commented:
Thank you @liancheng for the summary, which is clear for me who didn't dive into this part before. One thing that I am think about when I review the code #6833, how to remove the redundant files when user switch on the `speculative` in writing data via data source interface?

liancheng committed:
Uses job level UUID to take speculative tasks into account

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35258 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35258/consoleFull) for   PR 6864 at commit [`6d946bd`](https://github.com/apache/spark/commit/6d946bd5d3d4bc5b701d5ffa83e9d0934603faef).

liancheng commented:
@chenghao-intel Thanks for the comment! Speculation is a great point that I didn't notice. Updated this PR and now use a job level UUID instead of a task level one. Because essentially, what we want is to avoid name collision between different write jobs (potentially issued by different Spark applications). Within a single write job, we can always avoid name collision with the help of task ID.

SparkQA commented:
[Test build #35258 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35258/console) for   PR 6864 at commit [`6d946bd`](https://github.com/apache/spark/commit/6d946bd5d3d4bc5b701d5ffa83e9d0934603faef).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

liancheng committed:
Fixes Scala style issue

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35259 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35259/consoleFull) for   PR 6864 at commit [`14a47b9`](https://github.com/apache/spark/commit/14a47b90192be295f57f6f72a0f9e77a2e6b52b4).

liancheng commented:
@yhuai Updated PR description with an updated version of the summary commented above.  This is ready for review.

SparkQA commented:
[Test build #35259 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35259/console) for   PR 6864 at commit [`14a47b9`](https://github.com/apache/spark/commit/14a47b90192be295f57f6f72a0f9e77a2e6b52b4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/sources/commands.scala:
```diff
@@ -70,7 +71,7 @@ private[sql] case class InsertIntoHadoopFsRelation(
       s"Cannot write to multiple destinations: ${relation.paths.mkString(",")}")
 
-    val hadoopConf = sqlContext.sparkContext.hadoopConfiguration
+    val hadoopConf = new Configuration(sqlContext.sparkContext.hadoopConfiguration)
```
Do we need this? We already do `val job = new Job(hadoopConf)` below. BTW, we need to add comment to explain `new Job` will clone the conf.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/sources/commands.scala:
```diff
@@ -290,6 +298,9 @@ private[sql] abstract class BaseWriterContainer(
     setupConf()
 
+    ContextUtil.getConfiguration(job).set(
+      "spark.sql.sources.writeJobUUID", uniqueWriteJobId.toString)
```
Why we use a parquet method at here? Are we expecting that `getConfiguration` does not exist in some versions of `Job`?

yhuai replied:
We need to add comment to explain how this UUID get sent to executor side.

SparkQA commented:
[Test build #936 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/936/consoleFull) for   PR 6864 at commit [`14a47b9`](https://github.com/apache/spark/commit/14a47b90192be295f57f6f72a0f9e77a2e6b52b4).

SparkQA commented:
[Test build #936 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/936/console) for   PR 6864 at commit [`14a47b9`](https://github.com/apache/spark/commit/14a47b90192be295f57f6f72a0f9e77a2e6b52b4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

liancheng commented:
retest this please

liancheng commented:
Retesting for gaining more test failure logs to diagnose.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35313 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35313/consoleFull) for   PR 6864 at commit [`14a47b9`](https://github.com/apache/spark/commit/14a47b90192be295f57f6f72a0f9e77a2e6b52b4).

SparkQA commented:
[Test build #35313 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35313/console) for   PR 6864 at commit [`14a47b9`](https://github.com/apache/spark/commit/14a47b90192be295f57f6f72a0f9e77a2e6b52b4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

liancheng committed:
Adds more logs

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35344 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35344/consoleFull) for   PR 6864 at commit [`d5698b2`](https://github.com/apache/spark/commit/d5698b216c32a77633fa58d356ea2155659f68ba).

SparkQA commented:
[Test build #35344 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35344/console) for   PR 6864 at commit [`d5698b2`](https://github.com/apache/spark/commit/d5698b216c32a77633fa58d356ea2155659f68ba).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

lianhuiwang commented:
thanks @liancheng. @chenghao-intel there is other situation that needs to be considered when using data source interface.when some tasks are finished but job is failed because some tasks are failed, it needs to remove all output files of this job.

liancheng commented:
@lianhuiwang Yeah, thanks for reminding. We are also working on this issue. It will be addressed in another PR. At first, appending jobs with output committers like `DirectParquetOutputCommitter` can be tricky to handle since they writes directly to the target directory without using any temporary folder (this can be super useful for S3 since S3 file metadata operations and directory operations can be very slow). But with this PR, the job level UUID can be used to distinguish files written by different jobs.

liancheng commented:
With the help from @yhuai, finally found the root cause of the `OrcSourceSuite` failures showed in previous Jenkins builds. [SPARK-8501] [1] is opened to track that issue.

The reason why it shows in this PR and couldn't be reproduced locally on my laptop is that I changed the thread count number of the local `SparkContext` used by `TestHiveContext` to `*`, which uses 32 cores on Jenkins and 8 cores on my laptop. On the other hand, the testing data used in `OrcSourceSuite` consists of 10 rows, which means the ORC table written on my laptop consists of 8 part-files and each one contains some rows, while the one written on Jenkins consists of 32 part-files and some of them contains zero rows. It turned out that those empty ORC files messed things up. Please refer to [SPARK-8501] [1] for details.

For this reason, I made two more updates:

1. Change `local[*]` to `local[32]` for more determinism. 32 is chosen because Jenkins has 32 cores, and it should be enough for detecting concurrency issues.
2. Increased row number of the testing data used in `OrcSourceSuite` to 100 to temporarily workaround the build failure. SPARK-8501 will be fixed in another PR.

[1]: https://issues.apache.org/jira/browse/SPARK-8501

liancheng committed:
Works around SPARK-8501

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35361 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35361/consoleFull) for   PR 6864 at commit [`d412de7`](https://github.com/apache/spark/commit/d412de721b7d22a4f118e8b17dbc0faf5590e44f).

SparkQA commented:
[Test build #35361 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35361/console) for   PR 6864 at commit [`d412de7`](https://github.com/apache/spark/commit/d412de721b7d22a4f118e8b17dbc0faf5590e44f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala:
```diff
@@ -49,7 +49,7 @@ import scala.collection.JavaConversions._
   extends TestHiveContext(
     new SparkContext(
-      System.getProperty("spark.sql.test.master", "local[2]"),
+      System.getProperty("spark.sql.test.master", "local[32]"),
```
Maybe we should still use `local[*]?`

liancheng replied:
I think we'd better use a fixed number here to improve determinism (if we use 32 from the beginning, the ORC bug would be much easier to reproduce).

yhuai commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala:
```diff
@@ -44,7 +44,7 @@ abstract class OrcSuite extends QueryTest with BeforeAndAfterAll {
 
     sparkContext
-      .makeRDD(1 to 10)
+      .makeRDD(1 to 100)
```
Can we add a comment in this suite to document what's the problem we got (also create a jir)?

liancheng replied:
The JIRA is already there: https://issues.apache.org/jira/browse/SPARK-8501

Adding the comments.

yhuai commented on sql/hive/src/test/scala/org/apache/spark/sql/sources/SimpleTextRelation.scala:
```diff
@@ -156,6 +156,7 @@ class CommitFailureTestRelation(
         context: TaskAttemptContext): OutputWriter = {
       new SimpleTextOutputWriter(path, context) {
         override def close(): Unit = {
+          super.close()
```
Do we need this?

liancheng replied:
I was thinking about S3, where a file is not actually created before the output stream is closed (the `PUT` operation happens in `NativeS3FsOutputStream.close()`). But `SimpleTextRelation` is only used for local testing, so yeah, this line is not necessary.

liancheng replied:
I decided to leave it there. The writer should be closed anyway. Otherwise it's leaked.

SparkQA commented:
[Test build #950 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/950/consoleFull) for   PR 6864 at commit [`d412de7`](https://github.com/apache/spark/commit/d412de721b7d22a4f118e8b17dbc0faf5590e44f).

liancheng committed:
Uses job level UUID in SimpleTextRelation and avoids double task abortion

yhuai commented:
btw, do we need a PR for 1.4 backport?

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35416 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35416/consoleFull) for   PR 6864 at commit [`2738368`](https://github.com/apache/spark/commit/2738368eb04c256a4c28a802b5a3b12a5458fda8).

liancheng committed:
Adds comment about SPARK-8501

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35417 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35417/consoleFull) for   PR 6864 at commit [`3207323`](https://github.com/apache/spark/commit/32073239d8a03c6c7404c9dab7555c0cae1b5455).

SparkQA commented:
[Test build #950 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/950/console) for   PR 6864 at commit [`d412de7`](https://github.com/apache/spark/commit/d412de721b7d22a4f118e8b17dbc0faf5590e44f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #35416 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35416/console) for   PR 6864 at commit [`2738368`](https://github.com/apache/spark/commit/2738368eb04c256a4c28a802b5a3b12a5458fda8).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #35417 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35417/console) for   PR 6864 at commit [`3207323`](https://github.com/apache/spark/commit/32073239d8a03c6c7404c9dab7555c0cae1b5455).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

liancheng committed:
Workarounds SPARK-8513

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

liancheng commented:
It's proved that increasing thread number of the local `SparkContext` used by `TestHive` (and running the tests on a node with relatively more cores, say our Jenkins builder) is pretty useful for detecting concurrency related bugs.  SPARK-8501 and SPARK-8513 are both detected by this means.

SparkQA commented:
[Test build #35423 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35423/consoleFull) for   PR 6864 at commit [`99a73ab`](https://github.com/apache/spark/commit/99a73ab959ca95b41d6a69af8a48fea260d640fb).

SparkQA commented:
**[Test build #35423 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35423/console)**     for PR 6864 at commit [`99a73ab`](https://github.com/apache/spark/commit/99a73ab959ca95b41d6a69af8a48fea260d640fb)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

yhuai commented:
test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35429 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35429/consoleFull) for   PR 6864 at commit [`99a73ab`](https://github.com/apache/spark/commit/99a73ab959ca95b41d6a69af8a48fea260d640fb).

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/sources/commands.scala:
```diff
@@ -417,15 +428,14 @@ private[sql] class DefaultWriterContainer(
-        super.abortTask()
-        throw new RuntimeException("Failed to commit task", cause)
+    } catch { case cause: Throwable =>
+      throw new RuntimeException("Failed to commit task", cause)
```
This exception will be cached in `writeRows`, right? If so, can we add a comment and also explain how we will handle this exception?

yhuai replied:
Actually, I think we need to also add doc to `InsertIntoHadoopFsRelation` to explain the flow of this command and how we handle different kinds of failures/errors.

liancheng replied:
Right, it's handled in `writeRows`. Agree with more comments, I made multiple mistakes here myself...

yhuai commented:
LGTM. Left two comments regarding adding comments/docs.

liancheng committed:
Addresses comments

SparkQA commented:
[Test build #35429 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35429/console) for   PR 6864 at commit [`99a73ab`](https://github.com/apache/spark/commit/99a73ab959ca95b41d6a69af8a48fea260d640fb).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

liancheng committed:
More comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35437 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35437/consoleFull) for   PR 6864 at commit [`db7a46a`](https://github.com/apache/spark/commit/db7a46a169d1789ff221d7f84315edc3df04511a).

liancheng commented:
#6932 was opened to backport this PR to branch-1.4.

nemccarthy commented:
Can chance this can be merged today?

SparkQA commented:
[Test build #35437 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35437/console) for   PR 6864 at commit [`db7a46a`](https://github.com/apache/spark/commit/db7a46a169d1789ff221d7f84315edc3df04511a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

yhuai commented:
@nemccarthy Yeah. This one should be in today. I am taking a final check now.

yhuai commented:
LGTM. I am merging it to master.
